### PR TITLE
Add way to erase matter data only and emit event for application

### PR DESCRIPTION
**Pull request by jadhavrohit924**
- Problem
> We need a API to reset matter data ( chip config, counters and kvs)
> And event for the application about factory reset

- Changes
> Add events for factory reset and matter data reset
> Add InitiateMatterDataReset API to erase matter data

- Testing
> Tested factory reset and matter data reset with esp32

### CODE REVIEW

File: src/include/platform/CHIPDeviceEvent.h [hunk 1]
```diff
@@ -251,6 +251,16 @@ enum PublicEventTypes
...
      */
     kServerReady,
+
+    /**
+     * Signals the application that device is going to factory reset.
+     */
+    kFactoryReset,
+
+    /**
+     * Signals the application that device is going to reset matter data.
```
Comment (tcarmelveilleux) [inline src/include/platform/CHIPDeviceEvent.h:261]:
```suggestion
     * Signals the application that device is going to reset matter data (on last fabric removed).
```

File: src/include/platform/CHIPDeviceEvent.h [hunk 2]
```diff
@@ -251,6 +251,16 @@ enum PublicEventTypes
      * sending messages to other nodes.
      */
     kServerReady,
+
+    /**
+     * Signals the application that device is going to factory reset.
+     */
+    kFactoryReset,
+
+    /**
```
Comment (tcarmelveilleux) [inline src/include/platform/CHIPDeviceEvent.h:260]: It would be good to have plumbing (including unit-tested plumbing) in FabricTable to handle removal of last fabric, which could then, separately, emit this event from Server::Server.

Comment (jadhavrohit924) [reply src/include/platform/CHIPDeviceEvent.h:260]: Added way to emit event from Server

File: src/include/platform/ConfigurationManager.h [hunk 3]
```diff
@@ -129,8 +129,9 @@ class ConfigurationManager
     virtual void RunUnitTests() = 0;
 #endif
 
-    virtual bool IsFullyProvisioned()   = 0;
-    virtual void InitiateFactoryReset() = 0;
+    virtual bool IsFullyProvisioned()      = 0;
+    virtual void InitiateFactoryReset()    = 0;
+    virtual void InitiateMatterDataReset() = 0;
```
Comment (tcarmelveilleux) [inline src/include/platform/ConfigurationManager.h:134]: Please document what they are supposed to do.

File: src/app/server/Server.cpp [hunk 4]
```diff
@@ -515,11 +515,40 @@ void Server::ScheduleFactoryReset()
     PlatformMgr().ScheduleWork([](intptr_t) {
         // Delete all fabrics and emit Leave event.
         GetInstance().GetFabricTable().DeleteAllFabrics();
+        if (GetInstance().GetFabricTable().FabricCount() == 0)
```
Comment (tcarmelveilleux) [inline src/app/server/Server.cpp:518]: After DeleteAllFabrics, this will awlays be true. Why have the if?

File: src/app/server/Server.cpp [hunk 5]
```diff
@@ -515,11 +515,40 @@ void Server::ScheduleFactoryReset()
...
+            {
+                ChipLogError(DeviceLayer, "Failed to post factory reset event");
+            }
+        }
         PlatformMgr().HandleServerShuttingDown();
         ConfigurationMgr().InitiateFactoryReset();
     });
 }
 
+void Server::ScheduleMatterDataReset()
```
Comment (tcarmelveilleux) [inline src/app/server/Server.cpp:533]: This has a lot in common with "ScheduleFactoryReset". Why is it different? The existing `Server::ScheduleFactoryReset()` is really the "initiate matter data reset" with a name that is not the best. The kMatterDataReset event should be triggered from the Server's FabricTable::Delegate, when last fabric removed.

It would then flow:

- ScheduleFactoryReset
- ScheduleWork
  - DeleteAllFabrics()
    - <That's it>
- ServerFabricDelegate::OnFabricRemoved()
  - Check if 0 fabrics left. If so:
    - Post event (kMatterDataReset)
    - ScheduleWork of:
        - PlatformMgr().HandleServerShuttingDown()
        - ConfigurationMgr().InitiateMatterDataReset()

The overall trigger of "full factory reset" should be done internally by the product, if, after everything is deleted, and `kMatterDataReset` event is seen, the *product* decides to factory reset.

In other words, we don't need a `Server::ScheduleFactoryReset` since "factory reset" is never triggered inside of Matter --> only "all matter data" can be reset, and whether this causes a factory reset is a choice of the product and not part of the spec.

Comment (jadhavrohit924) [reply src/app/server/Server.cpp:533]: I agree to your point. Now, ScheduleFactoryReset will be replaced by ScheduleMatterReset and InitiateFactoryReset will be replaced by InitiateMatterReset
`ScheduleFactoryReset` ---------> `ScheduleMatterReset`
`InitiateFactoryReset`     ---------> `InitiateMatterReset`
which will erase matter data and `application will decide whether to erase additional data (Factory reset) or not.`

Since `matter reset` is mis-named to `factory reset`, some platforms deletes application data in matter reset. So with this change, is this expected that all the platforms need to update their matter reset logic?

**Comment by andy31415** [comment]
PR is old and has merge issues. It seems stale. Closing.
